### PR TITLE
[Diffusion] Fuse FLUX.2 token concatenation and NVFP4 quantization

### PR DESCRIPTION
**Pull request by BBuf**
## Summary

This PR fuses the FLUX.2 single-stream output sequence

```text
torch.cat([attention_bf16, swiglu_bf16]) -> FlashInfer NVFP4 quantize -> to_out GEMM
```

into one SM103 JIT kernel that reads both BF16 branches and writes the concatenated E2M1 values and 128x4-swizzled E4M3 scales directly. A standalone ModelOpt helper feeds the packed activations and scales directly to the existing FP4 `to_out` GEMM; this PR has no dependency on #37096.

This removes one 24,576-wide BF16 materialization and one full-width quantization read in each of the 48 single-stream blocks. Unsupported shapes, TP > 1, non-SM103 devices, compile/capture, alternate FP4 modes, or missing layout guarantees retain the existing path.

> [!NOTE]
> This PR is now based directly on current `main` and contains no #37096 commits. The benchmark below was originally collected with #37096 present in both baseline and candidate, so the reported delta still isolates only the token-cat fusion.

## Implementation

- add a lightweight JIT CUDA kernel using TensorRT-LLM's `cvt_warp_fp16_to_fp4`, the same conversion primitive used by FlashInfer;
- add a standalone prequantized-input ModelOpt GEMM helper so this PR does not depend on #37096's tuple interface;
- write packed values and swizzled scale factors in the exact layout expected by the existing NVFP4 GEMM;
- enable it only for FLUX.2-dev's TP1 `(6144 attention | 18432 MLP)` projection on SM103;
- keep eager BF16 cat + quantization as the fallback for torch.compile, CUDA graph capture, unsupported FP4 modes, shapes, devices, and row counts;
- fail clearly if the installed FlashInfer package does not provide the required NVFP4 headers.

## End-to-end benchmark

Hardware and workload:

- 1x NVIDIA GB300 288 GB, GPU clock locked to 1800 MHz
- `/scratch/models/FLUX.2-dev-NVFP4`
- native SGLang backend, all components resident
- 1024x1024, 50 denoising steps, guidance 4.0, seed 20260830, lossless PNG
- one request warmup per process
- 5 baseline + 5 PR processes in balanced order: `main, pr, pr, main, pr, main, main, pr, main, pr`
- historical isolated baseline: #37096 head `f37a2a874f`; candidate: `e423a737d3` (the same parent was present on both sides)

| Metric | Parent mean +/- stdev | PR mean +/- stdev | Reduction |
| --- | ---: | ---: | ---: |
| Denoise / step | 136.883 +/- 0.176 ms | 133.580 +/- 0.099 ms | **2.413%** |
| Denoise / request | 6844.131 +/- 8.819 ms | 6678.984 +/- 4.943 ms | **2.413%** |
| End-to-end / request | 6980.339 +/- 14.099 ms | 6813.233 +/- 4.094 ms | **2.394%** |

The committed microbenchmark measures `cat + flashinfer.fp4_quantize` against the fused producer:

| Tokens | Baseline | Fused | Speedup | Packed/scales exact |
| ---: | ---: | ---: | ---: | :---: |
| 17 | 68.78 us | 45.60 us | 1.51x | yes |
| 512 | 66.50 us | 44.53 us | 1.49x | yes |
| 4096 | 103.09 us | 43.95 us | 2.35x | yes |
| 4608 | 114.31 us | 47.70 us | 2.40x | yes |

## Profile evidence

A torch-profiler comparison of two denoise steps at 1024x1024 shows:

| Per denoise step | Parent | PR | Delta |
| --- | ---: | ---: | ---: |
| Total kernel launches | 1089 | 1041 | **-48** |
| Single-block BF16 cat launches | 48 | 0 | **-48** |
| Standalone NVFP4 quant launches | 104 | 56 | **-48** |
| Fused token-cat quant launches | 0 | 48 | +48 |
| Target cat + quant GPU time | 5464 us | 2261 us | **-58.625%** |
| Total GPU kernel duration | 129355 us | 126017 us | **-3337 us** |

### Isolated effect of this PR

All ten 50-step images in the 5+5 parent/PR run have the same SHA256:

```text
345079f4c6e1638499b3a55fd98d563d1ae71248efa0ef92723075066e7bab97
```

| Metric | Result |
| --- | ---: |
| Pixel exact | **yes** |
| SSIM | **1.000000** |
| PSNR | **infinity** |
| LPIPS (Alex) | **0.000000** |
| Pixel cosine similarity | **1.000000** |
| Max / mean absolute pixel difference | **0 / 0** |

The committed kernel test additionally requires the E2M1 bytes and swizzled E4M3 scale bytes to exactly match `flashinfer.fp4_quantize(torch.cat(...))`.

### Independent-branch accuracy scope

The token-cat producer is byte-exact against `flashinfer.fp4_quantize(torch.cat(...))`, and the isolated parent/candidate image comparison above is pixel-exact. The previously shown upstream-main vs complete-stack drift came from #37096, which is no longer part of this branch, so that comparison has been removed.

## Validation

Current independent head:

- `ruff check`, `ruff format --check`, `compileall`, and `git diff --check`: passed locally;
- diffusion import-surface suite: **9 passed** locally;
- a committed GPU test checks the standalone prequantized-input helper against the regular ModelOpt quantize + GEMM path for both default and TensorRT-LLM FlashInfer backends; GPU execution is delegated to CI because no SM103 development assignment is currently active.

Historical SM103 validation of the unchanged token-cat kernel at `e423a737d3`:

- exact GB300 kernel comparison against FlashInfer;
- native SGLang 1024x1024 eager smoke;
- `--enable-torch-compile --enable-breakable-cuda-graph` native FLUX.2 smoke;
- parent/candidate image comparison was pixel-exact.

## Reproduction

```bash
PYTHONPATH=python CUDA_VISIBLE_DEVICES=0 \
python test/manual/kernels/bench_flux2_token_cat_nvfp4.py
```

The full A/B runner, raw JSON summaries, profiler summary, and lossless comparison images are retained in the local validation artifact bundle.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #33386238626](https://github.com/sgl-project/sglang/actions/runs/33386238626)
Latest PR Test (Extra): :hourglass_flowing_sand: [Run #33386238330](https://github.com/sgl-project/sglang/actions/runs/33386238330)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #33386238640](https://github.com/sgl-project/sglang/actions/runs/33386238640)

### CODE REVIEW

**Comment by sushildubey171** [comment]
Hi @BBuf,  thanks for the work on this PR. I see minor difference in the posted image, not sure how above metric catches this differences, I have an accuracy metric for text-to-image models. If you're interested, it might be useful to run a pre-PR vs. post-PR accuracy comparison using: https://github.com/sgl-project/sglang/pull/25871

main-vs-pr_marked

**Comment by BBuf** [comment]
@sushildubey171 Thanks for flagging this. Your observation applied to the old comparison, but it no longer applies to the current version of this PR.

The visible difference in the old montage came from the stacked parent PR #37096, not from this token-concat fusion. I have now made #37141 independent of #37096; its current head (`e37c65c564`) is based directly on main (`afbca97f74`).

I revalidated the current standalone PR on an NVIDIA GB300 (SM103):

- The fused token-concat kernel produced exactly the same FP4 value bytes and scale bytes as `flashinfer.fp4_quantize(torch.cat(...))`.
- The standalone prequantized GEMM path exactly matched the regular quantize + GEMM path for both tested FlashInfer backends.
- For the same 1024x1024, 50-step, guidance 4.0, seed 20260830 end-to-end generation used in the comparison, base and PR produced byte-identical PNGs: SHA256 `1ec6d4d7102cb20cfc6dd2ae65c3932977006e26f1884e741c972f6e5174fecd`; pixel exact = true; max/mean pixel difference = 0.

Because the current change is byte-exact at the quantization boundary and pixel-exact end to end, an FID comparison would not add signal here. It would still be useful for a non-bit-exact optimization such as the now-separate #37096.

**Comment by BBuf** [comment]
One additional clarification about #37096: its non-bit-exact FLUX.2 NVFP4 FC1 + SwiGLU + FC2-input quantization fusion is now request-scoped and enabled only with `--quality=high`. The default `--quality=lossless` path keeps the original reference implementation.
